File: src/csharp/pulse/cdm/io/PBEnvironmentActions.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
using System.Collections.Generic;
namespace Pulse.CDM
{
  public class PBEnvironmentAction
  {

    #region Any Environment Action

    public static SEEnvironmentAction Load(pulse.cdm.bind.AnyEnvironmentActionData any)
    {
      if (any.ChangeEnvironmentalConditions != null)
      {
        SEChangeEnvironmentalConditions dst = new SEChangeEnvironmentalConditions();
        PBEnvironmentAction.Load(any.ChangeEnvironmentalConditions, dst);
        return dst;
      }
      if (any.ThermalApplication != null)
      {
        SEThermalApplication dst = new SEThermalApplication();
        PBEnvironmentAction.Load(any.ThermalApplication, dst);
        return dst;
      }
      return null;
    }

    public static pulse.cdm.bind.AnyEnvironmentActionData Unload(SEEnvironmentAction action)
    {
      pulse.cdm.bind.AnyEnvironmentActionData any = new pulse.cdm.bind.AnyEnvironmentActionData();
      if (action.GetType().IsAssignableFrom(typeof(SEChangeEnvironmentalConditions)))
      {
        any.ChangeEnvironmentalConditions = Unload((SEChangeEnvironmentalConditions)action);
        return any;
      }

      if (action.GetType().IsAssignableFrom(typeof(SEThermalApplication)))
      {
        any.ThermalApplication = Unload((SEThermalApplication)action);
        return any;
      }
      return null;
    }


    #endregion

    #region SEEnvironmentAction

    public static void Serialize(pulse.cdm.bind.EnvironmentActionData src, SEEnvironmentAction dst)
    {
      if (src.Action != null)
        PBAction.Serialize(src.Action, dst);
    }
    public static void Serialize(SEEnvironmentAction src, pulse.cdm.bind.EnvironmentActionData dst)
    {
      dst.Action = new pulse.cdm.bind.ActionData();
      PBAction.Serialize(src, dst.Action);
    }

    #endregion

    #region SEChangeEnvironmentalConditions

    public static void Load(pulse.cdm.bind.ChangeEnvironmentalConditionsData src, SEChangeEnvironmentalConditions dst)
    {
      dst.Clear();
      Serialize(src, dst);
    }
    public static void Serialize(pulse.cdm.bind.ChangeEnvironmentalConditionsData src, SEChangeEnvironmentalConditions dst)
    {
      if (src.EnvironmentAction != null)
        Serialize(src.EnvironmentAction, dst);
      if (!string.IsNullOrEmpty(src.EnvironmentalConditionsFile))
        dst.SetEnvironmentalConditionsFile(src.EnvironmentalConditionsFile);
      else if (src.EnvironmentalConditions != null)
        PBEnvironment.Load(src.EnvironmentalConditions, dst.GetEnvironmentalConditions());
    }

    public static pulse.cdm.bind.ChangeEnvironmentalConditionsData Unload(SEChangeEnvironmentalConditions src)
    {
      pulse.cdm.bind.ChangeEnvironmentalConditionsData dst = new pulse.cdm.bind.ChangeEnvironmentalConditionsData();
      Serialize(src, dst);
      return dst;
    }
    static void Serialize(SEChangeEnvironmentalConditions src, pulse.cdm.bind.ChangeEnvironmentalConditionsData dst)
    {
      dst.EnvironmentAction = new pulse.cdm.bind.EnvironmentActionData();
      Serialize(src, dst.EnvironmentAction);
      if (src.HasEnvironmentalConditionsFile())
        dst.EnvironmentalConditionsFile = src.GetEnvironmentalConditionsFile();
      else if (src.HasEnvironmentalConditions())
        dst.EnvironmentalConditions = PBEnvironment.Unload(src.GetEnvironmentalConditions());
    }

    #endregion

    #region SEThermalApplication

    public static void Load(pulse.cdm.bind.ThermalApplicationData src, SEThermalApplication dst)
    {
      dst.Clear();
      Serialize(src, dst);
    }
    public static void Serialize(pulse.cdm.bind.ThermalApplicationData src, SEThermalApplication dst)
    {
      if (src.EnvironmentAction != null)
        Serialize(src.EnvironmentAction, dst);
      if (src.ActiveHeating != null)
        PBEnvironment.Load(src.ActiveHeating, dst.GetActiveHeating());
      if (src.ActiveCooling != null)
        PBEnvironment.Load(src.ActiveCooling, dst.GetActiveCooling());
      if (src.AppliedTemperature != null)
        PBEnvironment.Load(src.AppliedTemperature, dst.GetAppliedTemperature());
    }

    public static pulse.cdm.bind.ThermalApplicationData Unload(SEThermalApplication src)
    {
      pulse.cdm.bind.ThermalApplicationData dst = new pulse.cdm.bind.ThermalApplicationData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(SEThermalApplication src, pulse.cdm.bind.ThermalApplicationData dst)
    {

      dst.EnvironmentAction = new pulse.cdm.bind.EnvironmentActionData();
      Serialize(src, dst.EnvironmentAction);
      dst.AppendToPrevious = !src.GetClearContents();

      if (src.HasActiveHeating())
        dst.ActiveHeating = PBEnvironment.Unload(src.GetActiveHeating());
      if (src.HasActiveCooling())
        dst.ActiveCooling = PBEnvironment.Unload(src.GetActiveCooling());
      if (src.HasAppliedTemperature())
        dst.AppliedTemperature = PBEnvironment.Unload(src.GetAppliedTemperature());
    }

    #endregion
  }
}

File: src/cpp/cdm/io/protobuf/PBEnvironmentActions.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
PUSH_PROTO_WARNINGS()
#include "pulse/cdm/bind/EnvironmentActions.pb.h"
POP_PROTO_WARNINGS()
#include "io/protobuf/PBEnvironmentActions.h"
#include "io/protobuf/PBEnvironment.h"
#include "io/protobuf/PBActions.h"
#include "system/environment/actions/SEChangeEnvironmentalConditions.h"
#include "system/environment/actions/SEThermalApplication.h"
#include "substance/SESubstanceManager.h"

void PBEnvironmentAction::Serialize(const CDM_BIND::EnvironmentActionData& src, SEEnvironmentAction& dst)
{
  PBAction::Serialize(src.action(), dst);
}
void PBEnvironmentAction::Serialize(const SEEnvironmentAction& src, CDM_BIND::EnvironmentActionData& dst)
{
  PBAction::Serialize(src, *dst.mutable_action());
}

void PBEnvironmentAction::Load(const CDM_BIND::ChangeEnvironmentalConditionsData& src, SEChangeEnvironmentalConditions& dst, const SESubstanceManager& subMgr)
{
  dst.Clear();
  PBEnvironmentAction::Serialize(src, dst, subMgr);
}
void PBEnvironmentAction::Serialize(const CDM_BIND::ChangeEnvironmentalConditionsData& src, SEChangeEnvironmentalConditions& dst, const SESubstanceManager& subMgr)
{
  PBEnvironmentAction::Serialize(src.environmentaction(), dst);
  if (!src.environmentalconditionsfile().empty())
    dst.SetEnvironmentalConditionsFile(src.environmentalconditionsfile());
  else if (src.has_environmentalconditions())
    PBEnvironment::Load(src.environmentalconditions(), dst.GetEnvironmentalConditions(), subMgr);
}
CDM_BIND::ChangeEnvironmentalConditionsData* PBEnvironmentAction::Unload(const SEChangeEnvironmentalConditions& src)
{
  CDM_BIND::ChangeEnvironmentalConditionsData* dst = new CDM_BIND::ChangeEnvironmentalConditionsData();
  PBEnvironmentAction::Serialize(src, *dst);
  return dst;
}
void PBEnvironmentAction::Serialize(const SEChangeEnvironmentalConditions& src, CDM_BIND::ChangeEnvironmentalConditionsData& dst)
{
  PBEnvironmentAction::Serialize(src, *dst.mutable_environmentaction());
  if (src.HasEnvironmentalConditionsFile())
    dst.set_environmentalconditionsfile(src.m_EnvironmentalConditionsFile);
  else if (src.HasEnvironmentalConditions())
    dst.set_allocated_environmentalconditions(PBEnvironment::Unload(*src.m_EnvironmentalConditions));
}
void PBEnvironmentAction::Copy(const SEChangeEnvironmentalConditions& src, SEChangeEnvironmentalConditions& dst, const SESubstanceManager& subMgr)
{
  dst.Clear();
  CDM_BIND::ChangeEnvironmentalConditionsData data;
  PBEnvironmentAction::Serialize(src, data);
  PBEnvironmentAction::Serialize(data, dst, subMgr);
}

void PBEnvironmentAction::Load(const CDM_BIND::ThermalApplicationData& src, SEThermalApplication& dst)
{
  dst.Clear();
  PBEnvironmentAction::Serialize(src, dst);
}
void PBEnvironmentAction::Serialize(const CDM_BIND::ThermalApplicationData& src, SEThermalApplication& dst)
{
  // Set this before our super class tells us to Clear if the action wants us to keep our current data
  dst.m_ClearContents = !src.appendtoprevious();
  PBEnvironmentAction::Serialize(src.environmentaction(), dst);
  if (src.has_activecooling())
    PBEnvironment::Load(src.activecooling(), dst.GetActiveCooling());
  if (src.has_activeheating())
    PBEnvironment::Load(src.activeheating(), dst.GetActiveHeating());
  if (src.has_appliedtemperature())
    PBEnvironment::Load(src.appliedtemperature(), dst.GetAppliedTemperature());
  dst.m_ClearContents = true;
}
CDM_BIND::ThermalApplicationData* PBEnvironmentAction::Unload(const SEThermalApplication& src)
{
  CDM_BIND::ThermalApplicationData* dst = new CDM_BIND::ThermalApplicationData();
  PBEnvironmentAction::Serialize(src, *dst);
  return dst;
}
void PBEnvironmentAction::Serialize(const SEThermalApplication& src, CDM_BIND::ThermalApplicationData& dst)
{
  PBEnvironmentAction::Serialize(src, *dst.mutable_environmentaction());
  if (src.HasActiveHeating())
    dst.set_allocated_activeheating(PBEnvironment::Unload(*src.m_ActiveHeating));
  if (src.HasActiveCooling())
    dst.set_allocated_activecooling(PBEnvironment::Unload(*src.m_ActiveCooling));
  if (src.HasAppliedTemperature())
    dst.set_allocated_appliedtemperature(PBEnvironment::Unload(*src.m_AppliedTemperature));
}
void PBEnvironmentAction::Copy(const SEThermalApplication& src, SEThermalApplication& dst)
{
  dst.Clear();
  CDM_BIND::ThermalApplicationData data;
  PBEnvironmentAction::Serialize(src, data);
  PBEnvironmentAction::Serialize(data, dst);
}

SEEnvironmentAction* PBEnvironmentAction::Load(const CDM_BIND::AnyEnvironmentActionData& any, const SESubstanceManager& subMgr)
{
  switch (any.Action_case())
  {
  case CDM_BIND::AnyEnvironmentActionData::ActionCase::kChangeEnvironmentalConditions:
  {
    SEChangeEnvironmentalConditions* a = new SEChangeEnvironmentalConditions(subMgr.GetLogger());
    PBEnvironmentAction::Load(any.changeenvironmentalconditions(), *a, subMgr);
    return a;
  }
  case CDM_BIND::AnyEnvironmentActionData::ActionCase::kThermalApplication:
  {
    SEThermalApplication* a = new SEThermalApplication();
    PBEnvironmentAction::Load(any.thermalapplication(), *a);
    return a;
  }
  }
  subMgr.Error("Unknown action type : " + any.Action_case());
  return nullptr;
}
CDM_BIND::AnyEnvironmentActionData* PBEnvironmentAction::Unload(const SEEnvironmentAction& action)
{
  CDM_BIND::AnyEnvironmentActionData* any = new CDM_BIND::AnyEnvironmentActionData();
  const SEChangeEnvironmentalConditions* cec = dynamic_cast<const SEChangeEnvironmentalConditions*>(&action);
  if (cec != nullptr)
  {
    any->set_allocated_changeenvironmentalconditions(PBEnvironmentAction::Unload(*cec));
    return any;
  }
  const SEThermalApplication* ta = dynamic_cast<const SEThermalApplication*>(&action);
  if (ta != nullptr)
  {
    any->set_allocated_thermalapplication(PBEnvironmentAction::Unload(*ta));
    return any;
  }
  delete any;
  return nullptr;
}